### DISCUSSION TITLE
How to do a many to many field

### DISCUSSION THREAD

**Discussion by snorresovold** [pocketbase/pocketbase#1740] [Q&A] [answered]
I have a database with users and products.

My products have a relation field relating to the users so i can get the usernames and such via 

`const product = await pb.collection('products').getOne(id, {
            expand: "user"
        })`

How would I do the same for users, and getting all products that a user has?
Should i add a relation field to my users?

**Comment by ganigeorgiev** [comment, marked as answer]
No, you don't have to add the reverse relation. Depending on your use case, you could:

- Filter the products for the user(s):
   ```js
   const products = await pb.collection("products").getFullList(200, { filter: `user = "${userId}"` })
   ```
   
- Use [indirect expand](https://pocketbase.io/docs/expanding-relations/#indirect-expand):
  ```js
  const user = await pb.collection("user").getOne("USER_ID", { expand: "products(user)") })

**Reply by snorresovold** [reply]
Getting some problems when running 

    export const load = async ({ params }: any) => {

    const fetchProduct = async (id: string) => {
        const user = await pb.collection("users").getOne(id, {
            expand: "products(user)" 
        })
        return user
    }

    return {
        user: await fetchProduct(params.userID)
    }}

On a Sveltekit App

Even though I get all the correct data I get an error when running 

`
{#each user.expand as product}
        <Product price={product.price} name={product.name} user={product.user} link={product.link}  desc={product.desc}/>
{/each}`



My +page.ts file looks like this

```
import { pb } from "$lib/pocketbase"

export const load = async ({ params }: any) => {

    const fetchProduct = async (id: string) => {
        const user = await pb.collection("users").getOne(id, {
            expand: "products(user)" 
        })
        return user
    }

    return {
        user: await fetchProduct(params.userID)
    }
}`

And my +page.svelte

```<script lang="ts">
	import Product from "$lib/Product.svelte";
    export let data: any;
    $: ({ user } = data);
    console.log(data)
</script>

<p>{user.username}</p>

<img
    class="avatar"
    src={`https://avatars.dicebear.com/api/identicon/${user.expand?.user?.username}.svg`}
    alt="avatar"
    width="40px"
/>


{#each user.expand as product}
    <Product price={product.price} name={product.name} user={product.user} link={product.link}  desc={product.desc}/>
{/each}
```

Here is also my data

```
  user: Record {
    avatar: '',
    collectionId: '_pb_users_auth_',
    collectionName: 'users',
    created: '2023-01-21 16:24:30.472Z',
    emailVisibility: false,
    expand: { 'products(user)': [Array] },
    id: 'jkn7egerj1ww5zq',
    name: '',
    updated: '2023-01-21 16:24:30.472Z',
    username: 'Espen',
    verified: false
  }`

**Reply by ganigeorgiev** [reply]
@snorresovold What is the error you are getting?

You can wrap with `try/catch` your await calls and inspect the error (to log the original response error - `console.log(err, err.originalError)`).

If the error is from the [autocancellation](https://github.com/pocketbase/js-sdk#auto-cancellation) you can either disable it (eg. globally via `pb.autoCancellation(false)`) or `try/catch` it and ignore `err.isAbort` (but keep in mind that the autocancellation usually indicates async bugs in user-land code and make sure that your component doesn't send excessive/unexpected requests):

```js
const fetchProduct = async (id: string) => {
    let user = {};

    try {
        user = await pb.collection("users").getOne(id, {
            expand: "products(user)" 
        })

    } catch (err) {
        if (!err.isAbort) {
            console.log("non-abort error...")
        }
    }
    
    return user
}
```

**Comment by snorresovold** [comment]
I keep getting a 500 server error on the svelte side because of the loop
```
{#each user.expand.user as product}
    <Product price={product.price} name={product.name} user={product.user} link={product.link}  desc={product.desc}/>
{/each}
```

When i delete it I can console log the data just fine.

Heres what te error looks like on the server
```
TypeError: Cannot read properties of undefined (reading 'length')
    at Module.each (/node_modules/svelte/internal/index.mjs:1857:31)
    at eval (/src/routes/users/[userID]/+page.svelte:22:25)
    at Object.$$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at Object.default (root.svelte:41:38)
    at eval (/src/routes/+layout.svelte:14:32)
    at Object.$$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at root.svelte:40:37
    at $$render (/node_modules/svelte/internal/index.mjs:1892:22)
    at Object.render (/node_modules/svelte/internal/index.mjs:1900:26)
    at Module.render_response (/node_modules/@sveltejs/kit/src/runtime/server/page/render.js:115:27)
```

**Reply by ganigeorgiev** [reply]
Are you sure that there is a `user.expand.user` field?
You can always `console.log(user)` to explore all available fields.

Based on your `fetchProduct()` method, the relation is indirect, so the expected field should be `user.expand['products(user)']` but make sure that it is array and that it is defined.

**Reply by snorresovold** [reply]
Thank you so much, it works now. I think the `user.expand['products(user)'] ` was the last piece of the puzzle.
